s5_subwords/readme.md

```diff
@@ -1,19 +1,50 @@
+ * [Unigram](https://arxiv.org/abs/1804.10959) 
+ * [Byte Pair Encoding](https://arxiv.org/abs/1508.07909)
+ * [Morfessor](https://morfessor.readthedocs.io/en/latest/)
+ 
+Please read about the tokenizer in their respective repositories, they all perform the same task, but the WER results will differ for each method. The default tokenizer is a "bpe", all scripts for that tokenizer are provided with this repository. To run the tokenizers "unigram" and the "sp_bpe" [Google's Sentence Piece](https://github.com/google/sentencepiece) is needed, a code snippet for a python virtual environment is provided along with a requirments.txt file. 
+
+For each subword tokenization method, there are two parameters, vocabulary size and boundary marking style. The vocabulary size is a parameter that controls how aggressively the words are tokenized, lower values mean fewer and smaller subword units and vice-versa for higher values. The boundary marking styles are the following (the sentence "dog walks" is used as an example): 
+* right-marked (r):  do+ g w+ alk+ s
+* left-marked (l):  do +g w +alk +s
+* left-right-marked (lr): do+ +g w+ +alk+ +s (The default) 
+* word-boundary (wb): do g + w alk s 
+
+The token used to mark the subword boundary can be any sign given that it's not a part of the language, in this setup the default is "+".
+
+### run/run_sw.sh
+The first step in the script is to train the subword tokenization model. Depending on its size, the text corpus or a subset of it is used for that. We have been using 4 million lines of text which should be more than enough. When the tokenizer is trained that model is used to tokenize the entire corpus. This subword tokenized text corpus will be used to prepare the Language model. There are three different subword tokenizers in the script Unigram, 
 
-Start by running the script "run_am.sh", it has all the steps related to the acoustic model training. It creates a LDA+MLLT+SAT system and then a TDNN model, with i-vectors and speed perturbation. The script is quite standard except for the creation of the dictionary as it uses the text corpus to create a grapheme based word-level lexicon. The word-level parts e.g. the "lang" and "dict" are called "base". Note that the text corpus must be normalized. That means one sentence per line, making all characters lower or upper case and removing all non-character related signs like numbers or punctuations. 
+The next step is to create the dictionary folder which that has the lexicon and other files that Kaldi needs. They are all made by extracting information from the subword tokenized text corpus. 
 
-# License
+The next step is to prepare the "lang" directory. For that, an external package is needed, which was developed by [Aalto University](https://github.com/aalto-speech/subword-kaldi.git). It is used to create the L.fst for the boundary marking style. Please read their paper, found in the repository, for more information. 
```

Comment:
can you add this repo as a git submodule?

```suggestion
The next step is to prepare the "lang" directory. For that, an external package, [subword-kaldi](https://github.com/aalto-speech/subword-kaldi.git) is needed, which was developed by [Aalto University](https://github.com/aalto-speech/subword-kaldi.git). It is used to create the L.fst for the boundary marking style. Please read their paper, found in the repository, for more information. 
```

Reply:
Will do